### DISCUSSION TITLE
Append Images to Gif

### DISCUSSION THREAD

**Discussion by Jay123210599** [ImageMagick/ImageMagick#7364] [Help]
### ImageMagick version

7.1.1-27

### Operating system, version and so on

Windows 11

### Description

I want to combine this single image to my gif (or video, if that's not possible) side by side. Meaning, I want to the image to remain on the left side of the gif for its entire duration (but not side the gif itself). How do I do that?
https://imgur.com/a/append-MC0jKn8

**Comment by fmw42** [comment]
So you want to overlay that single image onto every frame of the animation. Is that correct?  Your single image is much larger than the animation in width and height.  Do you want to resize the single image to fit one dimension?  If so, do you want to crop off excess or pad where it does not overlap?  Do you want the background white in the single image to be transparent when overlaying?  If not overlay, then where do you want to put the single image in relation to the animation?  Imagemagick can do any of the above. You just have to explain what you want!

Please clarify in detail how you want to combine the two images?

**Reply by Jay123210599** [reply]
"So you want to overlay that single image onto every frame of the animation. Is that correct?" Yes. "Do you want to resize the single image to fit one dimension?" Also yes. " If so, do you want to crop off excess or pad where it does not overlap? Do you want the background white in the single image to be transparent when overlaying?" Both no. " If not overlay, then where do you want to put the single image in relation to the animation?" Yes.

**Reply by fmw42** [reply]
Not a clear answer. You answered Yes to every part, but skipped one question.  Your Yes answers conflict in the last questions - Yes is not a proper answer to "where"?

**Reply by Jay123210599** [reply]
I want that image to be a part of every single frame in the animation, but I don't want it to block anything out of the gif. In other words, I want it to be on the left side of the gif.

**Reply by fmw42** [reply]
OK that is fine, but you said above you wanted it to be overlayed, not appended side-by-side.  

As I said below, I cannot open your supposed mp4 file. However, I can convert that to a gif animation from ffmpeg.  Is that OK before using Imagemagick to process your combination (append)?

**Reply by Jay123210599** [reply]
Yeah, that's fine. I also uploaded the actual gif file, if that helps.

Oh, and just so we're clear, I want them to be appended side-by-side, not overlayed.

**Reply by fmw42** [reply]
Your upload got changed to an mp4 again.  Please zip before uploading.

**Reply by Jay123210599** [reply]
Here it is.
https://drive.google.com/file/d/1QaBLAwHXcb4czwg5cQqazeFYeP1lxIDU/view?usp=sharing

**Comment by fmw42** [comment]
I am sorry, but your video file is rather strange.  When I try the following I get an error message. However, Mac QuickTime Player does play it.  I do not understand what is wrong.  I do have ffmpeg installed.  So perhaps it cannot handle that encoder.  Can you recreate in a more basic mp4 format?   Does any one else have the same error message. - what about @snibgo?

```
magick identify video.mp4
Unknown encoder 'webp'
```

Might this be a webp animation mislabeled as mp4?  Exiftool says it is mp4 with

```
0 MajorBrand                      : MP4 Base Media v1 [IS0 14496-12:2003]
2 CompressorID                    : avc1
- Encoder                         : Lavf58.12.100
```

**Comment by fmw42** [comment]
This is a Unix scripting solution. It will run in Linux, Mac OSX, Windows 10/11 _Unix_.  Offhand I cannot find a clever simple command to do it. Perhaps someone else has a clever way.

Create three directories: test1 and test2 and test3. Put your two input files into test1. Test2 and test3 will hold a number of temporary images that are the frames of the gif.



```
cd
cd desktop/test1

# resize webp image and save as png 
magick \( image.webp -resize x1080 \) image.png

# extract and save each frame of animated gif
magick output.gif +adjoin ../test2/out_%03d.gif

# do FOR loop setting the frames with each loop number added to 1000 so that have alphabetically sequential images
i=0
for img in ../test2/out_*.gif; do
j=$((i+1000))
echo "$i $j $img"
# append images and save in test3
magick image.png $img +append ../test3/anim_$j.gif
i=$((i+1))
done

# take all images in test3 and create the output animated gif
magick -dispose background -delay 10 ../test3/anim_*.gif -loop 0 animation.gif

```

<br>

http://www.fmwconcepts.com/misc_tests/animation.gif


The IM commands are common to Windows, but the FOR loop and variables have different styles.  Sorry, I do not know windows scripting.  So perhaps some kind Windows user can convert it for you.

**Reply by Jay123210599** [reply]
the output gif is slower than the input gif.

**Reply by fmw42** [reply]
Reduce the delay value in my command. You can go to 0, but some viewers put a minimum limit.

**Reply by Jay123210599** [reply]
What's the basic command line (not batch script) for this?

**Reply by fmw42** [reply]
As I said above, I cannot think of any way to do that that is just a simple command line.  Perhaps someone else has a better or simple solution.

**Comment by fmw42** [comment]
OK. I finally recalled how to do this. One has to splice a background image to animation on the left. Then layers composite the resized image onto spliced region.  Here is the Unix syntax command:

```
cd
cd desktop/test1
magick \
\( output.gif[0] -set option:height "%h" +delete \) \
\( image.webp -resize "x%[height]" -set option:newwidth "%w" -write mpr:img +delete \) \
\( output.gif -layers coalesce \
-background white -gravity west -splice "%[fx:newwidth]x0" \) \
NULL: \
mpr:img \
-gravity west -compose Over -layers composite \
-layers optimize \
animation2.gif
```

<br>

http://www.fmwconcepts.com/misc_tests/animation2.gif

For Windows remove the \ in front of the parentheses. You may also have to double your % to %% depending upon whether in a .bat file or not.

**Reply by Jay123210599** [reply]
Can you remove them for me?

**Reply by fmw42** [reply]
Remove what?

**Reply by Jay123210599** [reply]
the \ in front of the parentheses

**Comment by fmw42** [comment]
I do not guarantee this will work as I am not a Windows user. You also have to change the filenames and also the directories. If you put this in a .bat file, you will likely have to double the % to %%.

```
cd
cd desktop/test1
magick ^
( output.gif[0] -set option:height "%h" +delete ) ^
( image.webp -resize "x%[height]" -set option:newwidth "%w" -write mpr:img +delete ) ^
( output.gif -layers coalesce ^
-background white -gravity west -splice "%[fx:newwidth]x0" ) ^
NULL: ^
mpr:img ^
-gravity west -compose Over -layers composite ^
-layers optimize ^
animation2.gif

```

P.S.  Note that it is not fast for your example. On my Mac, it took about 2 minutes to finish.
